src/wallet/reserves.rs

```diff
@@ -0,0 +1,603 @@
+        psbt_in.sighash_type.is_some() && psbt_in.sighash_type != Some(SigHashType::All)
+    }) {
+        return Err(Error::ProofOfReservesInvalid(format!(
+            "Unsupported sighash type at input #{}",
+            i
+        )));
+    }
+
+    // Verify other inputs against prevouts and calculate the amount.
+    let serialized_tx = serialize(&tx);
+    if let Some((i, res)) = psbt
+        .inputs
+        .iter()
+        .zip(tx.input.iter())
+        .enumerate()
+        .map(|(i, (psbt_in, tx_in))| {
+            if let Some(utxo) = &psbt_in.witness_utxo {
+                (i, &utxo.script_pubkey, utxo.value)
+            } else if let Some(nwtx) = &psbt_in.non_witness_utxo {
+                let outp = &nwtx.output[tx_in.previous_output.vout as usize];
+                (i, &outp.script_pubkey, outp.value)
+            } else {
+                panic!("must be either witness or legacy");
+            }
+        })
+        .map(|(i, script, value)| {
+            (
+                i,
+                bitcoinconsensus::verify(script.to_bytes().as_slice(), value, &serialized_tx, i),
+            )
+        })
+        .find(|(_i, res)| res.is_err())
+    {
+        return Err(Error::ProofOfReservesInvalid(format!(
+            "Signature validation error at input #{}: {:?}",
+            i,
+            res.err().unwrap()
+        )));
+    }
+
+    // calculate the spendable amount of the proof
+    let sum = tx
+        .input
+        .iter()
+        .zip(psbt.inputs.iter())
+        .map(|(tx_in, psbt_in)| {
+            if let Some(utxo) = &psbt_in.witness_utxo {
+                utxo.value
+            } else if let Some(nwtx) = &psbt_in.non_witness_utxo {
+                nwtx.output[tx_in.previous_output.vout as usize].value
+            } else {
+                0
+            }
+        })
+        .fold(0, |acc, val| acc + val);
+
+    // verify the unspendable output
+    let pkh = PubkeyHash::from_hash(hash160::Hash::hash(&[0]));
+    let out_script_unspendable = bitcoin::Address {
+        payload: Payload::PubkeyHash(pkh),
+        network: network,
+    }
+    .script_pubkey();
+    if tx.output[0].script_pubkey != out_script_unspendable {
+        return Err(Error::ProofOfReservesInvalid("Invalid output".to_string()));
+    }
+
+    // inflow and outflow being equal means no miner fee
+    if tx.output[0].value != sum {
+        return Err(Error::ProofOfReservesInvalid(
+            "In and out values are not equal".to_string(),
+        ));
+    }
+
+    Ok(sum)
+}
+
+/// Construct a challenge input with the message
+fn challenge_txin(message: &str) -> TxIn {
+    let message = "Proof-of-Reserves: ".to_string() + message;
+    let message = sha256d::Hash::hash(message.as_bytes());
+    TxIn {
+        previous_output: OutPoint::new(Txid::from_hash(message), 0),
+        sequence: 0xFFFFFFFF,
+        script_sig: Builder::new().into_script(),
+        witness: Vec::new(),
```

Comment:
You can probably replace those two lines with:

```suggestion
        ..Default::default(),
```